src/matrix.rs

```diff
@@ -85,9 +89,90 @@ macro_rules! define_square_matrix_struct {
     };
 }
 
+macro_rules! impl_sub_matrix {
+    ($parent:ident, $child:ident) => {
+        impl $parent {
+            fn submatrix(&self, row: usize, col: usize) -> $child {
+                let mut res = $child::zero();
+                for i in 0..$parent::order() {
+                    if i != row {
+                        let k = if i < row { 0 } else { 1 };
+                        for j in 0..$parent::order() {
+                            if j != col {
+                                let l = if j < col { 0 } else { 1 };
+                                res[i - k][j - l] = self[i][j];
+                            }
+                        }
+                    }
+                }
+                res
+            }
+        }
+    };
+}
+
+macro_rules! impl_determinant {
+    ($parent:ident) => {
+        impl $parent {
+            fn minor(&self, row: usize, col: usize) -> f64 {
+                self.submatrix(row, col).determinant()
+            }
+
+            fn cofactor(&self, row: usize, col: usize) -> f64 {
+                let minor = self.minor(row, col);
+                if (row + col) % 2 == 0 {
+                    minor
+                } else {
+                    -minor
+                }
+            }
+
+            fn determinant(&self) -> f64 {
+                let mut res = 0.0;
+                for i in 0..Self::order() {
+                    res += (self[0][i] * self.cofactor(0, i));
+                }
+                res
+            }
+        }
+    };
+}
+
 define_square_matrix_struct!(Mat2x2, 2);
 define_square_matrix_struct!(Mat3x3, 3);
 define_square_matrix_struct!(Mat4x4, 4);
+impl_sub_matrix!(Mat4x4, Mat3x3);
+impl_sub_matrix!(Mat3x3, Mat2x2);
+impl_determinant!(Mat3x3);
+impl_determinant!(Mat4x4);
+
+impl Mat2x2 {
+    fn determinant(&self) -> f64 {
+        self[0][0] * self[1][1] - self[0][1] * self[1][0]
+    }
+}
+
+impl Mat4x4 {
+    fn invertible(&self) -> bool {
+        self.determinant().abs() > std::f64::EPSILON
+    }
+
+    fn inverse(&self) -> Result<Self, &str> {
+        if self.invertible() {
+            let mut res = Mat4x4::zero();
+            let det = self.determinant();
+            for r in 0..Self::order() {
+                for c in 0..Self::order() {
+                    let cof = self.cofactor(r, c);
+                    res[c][r] = cof / det;
+                }
+            }
+            Ok(res)
+        } else {
+            Err("Matrix is not invertible")
```

Comment:
Is it possible to have a simple test for this case also?

Reply:
Will fix!